### PR TITLE
Build cgnstools and OpenGL

### PR DESCRIPTION
**Pull request by brtnfld**
Place holder for discussion and continuation of PR #178.

### CODE REVIEW

**Comment by brtnfld** [comment]
I tried on my mac, and during cmake configure:

-- Found Tclsh: /usr/bin/tclsh (found version "8.5")
-- Found TCL: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/tcl.framework
-- Found TCLTK: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/tcl.framework
-- Found TK: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/tk.framework
-- Found OpenGL: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework

But it then failed in building:

[  2%] Building Fortran object src/CMakeFiles/cgns_shared.dir/cgns_f.F90.o
In file included from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/get.c:8:
/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:5:10: fatal error: OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h: No such file or directory
    5 | #include "OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/cgnstools/tkogl/CMakeFiles/tkogl.dir/get.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[  2%] Building C object src/CMakeFiles/cgns_shared.dir/cgns_error.c.o
In file included from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.h:10,
                 from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/gencyl.c:13:
/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:5:10: fatal error: OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h: No such file or directory
    5 | #include "OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.h:10,
                 from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.c:14:
/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:5:10: fatal error: OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h: No such file or directory
    5 | #include "OPENGL_INCLUDE_SUBDIR-NOTFOUND/gl.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/cgnstools/tkogl/CMakeFiles/tkogl.dir/gencyl.c.o] Error 1
make[2]: *** [src/cgnstools/tkogl/CMakeFiles/tkogl.dir/tkogl.c.o] Error 1
make[1]: *** [src/cgnstools/tkogl/CMakeFiles/tkogl.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

File: src/cgnstools/CMakeLists.txt [hunk 1]
```diff
@@ -95,3 +95,10 @@ CG_LIB_DIR=${CMAKE_INSTALL_PREFIX}/share/cgnstools; export CG_LIB_DIR
 	${CMAKE_CURRENT_BINARY_DIR}/cgconfig
 	DESTINATION bin)
 endif (WIN32)
+
+find_file(OPENGL_INCLUDE_SUBDIR
+	NAMES GL OpenGL
```
Comment (MicK7) [inline src/cgnstools/CMakeLists.txt]:
```suggestion
	NAMES GL OpenGL .
```

**Comment by brtnfld** [comment]
It looks like the complete path is not set.

/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:5:10: fatal error: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/./gl.h: No such file or directory
    5 | #include "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/./gl.h"
 
The path to the headers are:

ls /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Versions/A/Headers/
total 344
drwxr-xr-x  21 root   672B Jan 14 11:20 ./
drwxr-xr-x   6 root   192B Jan 14 11:20 ../
-rw-r--r--   1 root   557B Oct  2 01:06 CGLContext.h
-rw-r--r--   1 root   467B Oct  2 01:06 CGLCurrent.h
-rw-r--r--   1 root   980B Oct  2 01:06 CGLDevice.h
-rw-r--r--   1 root    14K Oct  2 01:06 CGLIOSurface.h
-rw-r--r--   1 root   131K Oct  2 01:06 CGLMacro.h
-rw-r--r--   1 root   3.1K Oct  2 01:06 CGLRenderers.h
-rw-r--r--   1 root    21K Oct  2 01:06 CGLTypes.h
-rw-r--r--   1 root   6.6K Oct  2 01:06 OpenGL.h
-rw-r--r--   1 root   775B Oct  2 01:06 OpenGLAvailability.h
-rw-r--r--   1 root   167K Oct  2 01:06 gl.h
-rw-r--r--   1 root   178K Oct  2 01:06 gl3.h
-rw-r--r--   1 root    14K Oct  2 01:06 gl3ext.h
-rw-r--r--   1 root   212K Oct  2 01:06 glext.h
-rw-r--r--   1 root   327B Oct  2 01:06 gliContext.h
-rw-r--r--   1 root    86K Oct  2 01:06 gliDispatch.h
-rw-r--r--   1 root   1.4K Oct  2 01:06 gltypes.h
-rw-r--r--   1 root    18K Oct  2 01:06 glu.h
-rw-r--r--   1 root   3.8K Oct  2 01:06 gluContext.h
-rw-r--r--   1 root   2.8K Oct  2 01:06 gluMacro.h

**Comment by brtnfld** [comment]
/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:7:10: fatal error: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers/glx.h: No such file or directory
    7 | #include "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers/glx.h" 

framework does not have glx.h

From cmake docs:

> macOS-Specific
> ^^^^^^^^^^^^^^
> On OSX FindOpenGL defaults to using the framework version of OpenGL. People
> will have to change the cache values of OPENGL_glu_LIBRARY and
> OPENGL_gl_LIBRARY to use OpenGL with X11 on OSX.

.//System/Volumes/Data/opt/local/include/xcb/glx.h
.//System/Volumes/Data/opt/X11/include/xcb/glx.h
.//System/Volumes/Data/opt/X11/include/GL/glx.h

File: src/cgnstools/common/gl_config.h.in [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#ifdef _WIN32
+#include <GL/gl.h>
+#include <GL/glu.h>
+#else
+#include "@OPENGL_INCLUDE_SUBDIR@/gl.h"
+#include "@OPENGL_INCLUDE_SUBDIR@/glu.h"
+#include "@OPENGL_INCLUDE_SUBDIR@/glx.h"
```
Comment (MicK7) [inline src/cgnstools/common/gl_config.h.in]:
```suggestion
#include <GL/glx.h>
```

**Comment by brtnfld** [comment]
Building C object src/cgnstools/tkogl/CMakeFiles/tkogl.dir/tkogl.c.o
cd /Users/brtnfld/packages/CGNS/build/src/cgnstools/tkogl && /Users/brtnfld/packages/mpich/bin/mpicc -DH5_BUILT_AS_STATIC_LIB -I/Users/brtnfld/packages/CGNS/build/src -I/Users/brtnfld/packages/CGNS/src -I/Users/brtnfld/packages/hdf5/build/hdf5/include -I/Users/brtnfld/packages/zlib-1.2.11/zlib/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tcl.framework/Headers -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/System/Library/Frameworks/Tk.framework/Headers -I/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common  -g -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk   -o CMakeFiles/tkogl.dir/tkogl.c.o   -c /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.c
In file included from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.h:10,
                 from /Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/tkogl.c:14:
/Users/brtnfld/packages/CGNS/src/cgnstools/tkogl/../common/gl_config.h:7:10: fatal error: GL/glx.h: No such file or directory
    7 | #include <GL/glx.h>
      |          ^~~~~~~~~~
compilation terminated.

**Comment by gsjaardema** [comment]
I get the same missing GL/glx.h include error as Scot.  On my OSX system, the `GL/glx.h` include file is in `/opt/X11/include`

**Comment by brtnfld** [comment]
Yes, cmake:

find_package(GLUT   REQUIRED)
find_package(OpenGL REQUIRED)

finds the frameworks packages. It needs to find the xquartz installs

**Comment by brtnfld** [comment]
I think the solution is to use 
find_package(X11 REQUIRED)

**Comment by brtnfld** [comment]
Everything compiles now on OSX, but I'm having an issue running cgnsview.

dyld: Library not loaded: /Users/brtnfld/packages/mpich-3.2.1/../mpich/lib/libmpi.12.dylib
  Referenced from: /usr/local/lib/libz.1.dylib
  Reason: no suitable image found.  Did find:

But it might be an issue with my set-up. I'll look into it. 

Also, the install cgnsview wrapper script  reports an error:

./cgnsview
Error: cgnsview.tcl script not found

Probably related to https://cgnsorg.atlassian.net/browse/CGNS-125

@gsjaardema does cgnsview work for you?

**Comment by gsjaardema** [comment]
@brtnfld  The build and install seems to work correctly for me.  cgnsview works when installed; however, it do not get any labels on the buttons.  See image below:

image

**Comment by brtnfld** [comment]
@gsjaardema, I fixed my issue, but all that I get is a blank white screen.

**Comment by brtnfld** [comment]
On windows, cmake finds the library correctly:

-- Found Tclsh: C:/ActiveTcl/bin/tclsh.exe (found version "8.6")
-- Found TCL: C:/ActiveTcl/lib/tcl86t.lib
-- Found TCLTK: C:/ActiveTcl/lib/tcl86t.lib
-- Found TK: C:/ActiveTcl/lib/tk86t.lib

but I think there is an issue with  Intels compiler:
 

[ 68%] Building C object src/cgnstools/cgnsview/CMakeFiles/cgiowish.dir/winmain.c.obj
        cd C:\Users\Public\Downloads\IntelBuild_parallel\build_cgns\src\cgnstools\cgnsview
        C:\PROGRA~2\INTELS~1\COMPIL~1.281\windows\bin\intel64\icl.exe @C:\Users\SCOTBR~1\AppData\Local\Temp\nm862B.tmp
icl -DUSE_HTMLHELP -DWIN32 -D_CRT_SECURE_NO_WARNINGS -D_HDF5USEDLL_ -IC:\Users\Public\Downloads\IntelBuild_parallel\build_cgns\src -IC:\Users\Public\Downloads\IntelBuild_parallel\CGNS\src -IC:\PROGRA~2\INTELS~1\COMPIL~1.281\windows\mpi\intel64\include -IC:\ActiveTcl\include -IC:\Users\Public\Downloads\IntelBuild_parallel\CGNS\src\cgnstools\cgnsview\..\common -IC:\Users\Public\Downloads\IntelBuild_parallel\CGNS\src\cgnstools\cgnsview\..\.. -IC:\PROGRA~2\WI3CF2~1\10\Include\100177~1.0\um -IC:\PROGRA~1\HDF_GR~1\HDF5\113~1.0MP\include -DUPPERCASE /MD /Qstd=c99 -DNDEBUG /MD /O2 /Ob2 /DNDEBUG /FoCMakeFiles\cgiowish.dir\winmain.c.obj /FdCMakeFiles\cgiowish.dir\ -c C:\Users\Public\Downloads\IntelBuild_parallel\CGNS\src\cgnstools\cgnsview\winmain.c

Intel(R) C++ Intel(R) 64 Compiler for applications running on Intel(R) 64, Version 19.0.5.281 Build 20190815
Copyright (C) 1985-2019 Intel Corporation.  All rights reserved.

winmain.c
C:\PROGRA~2\WI3CF2~1\10\Include\100177~1.0\um\winnt.h(20580): error: expected an identifier
          const VOID* None;
                      ^

C:\PROGRA~2\WI3CF2~1\10\Include\100177~1.0\um\processthreadsapi.h(1032): error: expected an identifier
      ULONG ControlMask;
            ^

C:\PROGRA~2\WI3CF2~1\10\Include\100177~1.0\um\processthreadsapi.h(1151): error: expected an identifier
      ULONG ControlMask;
            ^

**Comment by brtnfld** [comment]
I think we should merge it as is. It at least works on linux, and we are now test building cgnstools. We need to get a patched version out soon, and at least cgnstools compiles.

It looks like these other issues on OSX and Windows might be more involved than a simple patch. We can list them as known issues in the release notes. Thoughts?

**Comment by MicK7** [comment]
I agree with you. OSX and Windows issue might involved more than a patch (ideally switching to EGL + OpenGL2)  instead of glx+OpenGL1 , this would allow future handling of wayland, using Framework on OSX ...)

**Comment by gsjaardema** [comment]
I did discover that if I deselect 'show toolbar' and then select it again, then all the labels and icons show up in my Mac cgnsview.

image

**Pull request merged**